Update after Erik's feedback (#1)
rchiodo/peps: pep-0722.rst

PEP: 722
Title: Stricter Type Guards

> Author: Rich Chiodo <rchiodo at microsoft.com>, Eric Traut <erictr at microsoft.com>, Erik De Bonte <erikd at microsoft.com>

Reviewer: You should use a non-microsoft email address for Eric.

Author: Unfortunately building crashes if that's the case. The pep build system only allows one e-mail per name. So I'd have to update all of the other PEPs that Eric is mentioned in.

> Author: Rich Chiodo <rchiodo at microsoft.com>, Eric Traut <erictr at microsoft.com>, Erik De Bonte <erikd at microsoft.com>

Reviewer: Not sure I deserve to be in the authors list. Your call. You'll get feedback from lots of people and probably don't want to add all feedback providers to this list.

Author: Hmm, okay. You are the first person to look at it though :)

Abstract
========

:pep:`647` introduced the concept of ``TypeGuard`` functions which return true
if their input parameter matches their target type. For example, a function that
returns ``TypeGuard[str]`` is assumed to return ``true`` if and only if it's
input parameter is a ``str``. This allows type checkers to narrow types in this
positive case.

> This PEP further refines :pep:`647` by allowing type checkers to narrow types
> even further when a ``TypeGuard`` function returns false.

Reviewer: ```suggestion
This PEP further refines :pep:`647` by allowing type checkers to also narrow types
when a ``TypeGuard`` function returns false.
```

Motivation
==========

`TypeGuards <typeguards_>`__ are used throughout Python libraries to allow a
type checker to narrow the type of something when the ``TypeGuard``
returns true.

.. code-block:: python

    def is_str(val: str | int) -> TypeGuard[str]:
        return isinstance(val, str)

    def func(val: str | int):
        if is_str(val):
            # Type checkers can assume val is a 'str' in this branch

However, in the ``else`` clause, :pep:`647` didn't prescribe what the type might
be:

.. code-block:: python

    def is_str(val: str | int) -> TypeGuard[str]:
        return isinstance(val, str)

    def func(val: str | int):
        if is_str(val):
            # Type checkers can assume val is a 'str' in this branch
        else:
            # Type here is not narrowed. It is still 'str | int' 

This PEP proposes that when the type argument of the ``TypeGuard`` is a subtype

> of the type of the first input argument, then the false case can be further

Reviewer: "first input argument" -- are type guards allowed to have more than one parameter?

Author: They are. The typechecker only looks at the first one though.

> of the type of the first input argument, then the false case can be further

Reviewer: IIRC, an argument is a value that you pass to a function, so what you're talking about here is a parameter.

```suggestion
of the type of the first input parameter, then the false case can be further
```

be:

.. code-block:: python

    def is_str(val: str | int) -> TypeGuard[str]:
        return isinstance(val, str)

    def func(val: str | int):
        if is_str(val):
            # Type checkers can assume val is a 'str' in this branch
        else:
            # Type here is not narrowed. It is still 'str | int' 

This PEP proposes that when the type argument of the ``TypeGuard`` is a subtype
of the type of the first input argument, then the false case can be further
narrowed. 

This changes the example above like so:

.. code-block:: python

    def is_str(val: str | int) -> TypeGuard[str]:
        return isinstance(val, str)

    def func(val: str | int):
        if is_str(val):
            # Type checkers can assume val is a 'str' in this branch
        else:
            # Type checkers can assume val is an 'int' in this branch

Since the ``TypeGuard`` type (or output type) is a subtype of the input argument
type, a type checker can determine that the only possible type in the ``else``
is the other type in the Union. In this example, it is safe to assume that if
``is_str`` returns false, then type of the ``val`` argument is an ``int``.

> Unsafe Narrowing

Reviewer: It seems weird to me that the "Unsafe narrowing" and "Creating invalid narrowing" sections are within "Motivation"

Author: Where do you think they should go? Rejected ideas?

Author: In Eric's PEP he does describe some of the situations where TypeGuards don't work in the motivation section:
https://peps.python.org/pep-0647/

Reviewer: I was thinking "Specification". Counter examples don't strike me as part of the *motivation* for this PEP.

Author: Sounds good.

--------------------

There are cases where this further type narrowing is not possible. Here's an
example:

.. code-block:: python

    def is_str_list(val: list[int | str]) -> TypeGuard[list[str]]
        return all(isinstance(x, str) for x in val)

    def func(val: list[int | str]):
        if is_str_list(val):
            # Type checker assumes list[str] here
        else:
            # Type checker cannot assume list[int] here

> Since ``list`` is invariant, it doesn't have any subtypes. This means type
> checkers cannot narrow the type to ``list[int]`` in the false case.
> ``list[str]`` is not a subtype of ``list[str | int]``.

Reviewer: Suggest moving the last sentence after the first sentence.

```suggestion
Since ``list`` is invariant, it doesn't have any subtypes. ``list[str]`` is not a subtype of ``list[str | int]``. This means type
checkers cannot narrow the type to ``list[int]`` in the false case.
```

Type checkers should not assume any narrowing in the false case when the

> ``TypeGuard`` type is not a subtype of the input argument type. 

Reviewer: Maybe "TypeGuard type argument" is a good way to refer to the "output type"?

```suggestion
``TypeGuard`` type argument is not a subtype of the input argument type. 
```

type checker can assume the list is a ``list[str]`` if the ``TypeGuard``
function returns true.

Creating invalid narrowing
--------------------------

The new ``else`` case for a ``TypeGuard`` can be setup incorrectly. Here's an example:

.. code-block:: python

    def is_positive_int(val: int | str) -> TypeGuard[int]:
        return isinstance(val, int) and val > 0

    def func(val: int | str):
        if is_positive_int(val):
            # Type checker assumes int here
        else:
            # Type checker assumes str here

A type checker will assume for the else case that the value is ``str``. This
is a change in behavior from :pep:`647` but as that pep stated `here <https://peps.python.org/pep-0647/#enforcing-strict-narrowing>`__
there are many ways a determined or uninformed developer can subvert type safety.

A better way to handle this example would be something like so:

.. code-block:: python

    PosInt = NewType('PosInt', int)

    def is_positive_int(val: PosInt | int | str) -> TypeGuard[PosInt]:
        return isinstance(val, int) and val > 0

    def func(val: int | str):
        if is_positive_int(val):

>             # Type checker assumes PosInt here

Reviewer: > Type checker assumes PosInt here

This is true even though `val` is `int | str` (i.e. it doesn't include `PosInt`)?

Author: Yes that's true. This works because PosInt is an int.

Author: You should be able to try it yourself in the latest pylance.

Create a pyrightconfig.json with this in it:

```json
{
    "enableExperimentalFeatures" : true
}
```

Then the TypeGuard will behave as I described here.

Reviewer: I believe that that's how Pyright works. I guess I'm wondering:

a) Is this correct?

b) What would be the behavior if `PosInt` was unrelated to any of the types in the `val` union? Consider the following example. Within the `if` block, is `val` considered to be `datetime` because `is_datetime` returned `true`? Or is it `Never` because the intersection of `datetime` and `int | str` is empty? Oh, I guess this might be the multiple inheritance thing again. Maybe the `val` passed into `func` is derived from both `int` and `datetime`?

```python
def is_datetime(val: datetime | int | str) -> TypeGuard[datetime]:
    return isinstance(val, datetime)

def func(val: int | str):
    if is_datetime(val):
        # Is val a datetime here or Never?
```

c) Should the PEP call more attention to these issues?

Author: I believe the TypeGuard pep says that if the TypeGuard returns something nonsensical, that's the user's fault. It will just happily presume that's the actual type if the TypeGuard returns true.

This is exactly what Pyright does at the moment.

Author: It's mentioned here:
https://peps.python.org/pep-0647/#enforcing-strict-narrowing

Abstract
========

:pep:`647` created a special return type annotation ``TypeGuard`` that allowed

> type checkers to narrow types.

Reviewer: I think a little more context would still be helpful here.

```suggestion
:pep:`647` introduced the concept of ``TypeGuard`` functions which return true if their input parameter matches their target type. For example, a function that returns ``TypeGuard[str]`` is assumed to return ``true`` if and only if it's input parameter is a ``str``. This allows type checkers to narrow types in this positive case.
```

This PEP further refines :pep:`647` by allowing type checkers to narrow types
even further when a ``TypeGuard`` function returns false.

Motivation
==========

`TypeGuards <typeguards_>`__ are used throughout Python libraries to allow a

> type checker to narrow what the type of something is when the ``TypeGuard``

Reviewer: ```suggestion
type checker to narrow the type of something when the ``TypeGuard``
```

returns true.

.. code-block:: python

    def is_str(val: str | int) -> TypeGuard[str]:
        return isinstance(val, str)

    def func(val: str | int):
        if is_str(val):

>             # Type checkers can assume val is a 'str' in this branch

Reviewer: Maybe this first code block could be removed since the second one is identical except for adding the `else` block?

Author: I think it should be there cause I didn't want to introduce the else case until after the user had seen the if case.

Reviewer: Just want to make sure we're talking about the same thing. I was envisioning the following. And maybe to mention in the `if` block comment that the "val is a str" behavior is already defined by PEP 647.

```text
`TypeGuards` are used throughout Python libraries to allow a
type checker to narrow the type of something when the ``TypeGuard``
returns true. However, in the ``else`` clause, :pep:`647` didn't prescribe
what the type might be:
```

```python
    def is_str(val: str | int) -> TypeGuard[str]:
        return isinstance(val, str)

    def func(val: str | int):
        if is_str(val):
            # Type checkers can assume val is a 'str' in this branch
        else:
            # Type here is not narrowed. It is still 'str | int' 
```

However, in the ``else`` clause, :pep:`647` didn't prescribe what the type might
be:

.. code-block:: python

    def is_str(val: str | int) -> TypeGuard[str]:
        return isinstance(val, str)

    def func(val: str | int):
        if is_str(val):
            # Type checkers can assume val is a 'str' in this branch
        else:
            # Type here is not narrowed. It is still 'str | int' 

This PEP proposes that when the type argument of the ``TypeGuard`` is a subtype
of the type of the first input argument, then the false case can be further
narrowed. 

This changes the example above like so:

.. code-block:: python

    def is_str(val: str | int) -> TypeGuard[str]:
        return isinstance(val, str)

    def func(val: str | int):
        if is_str(val):
            # Type checkers can assume val is a 'str' in this branch
        else:

>             # Type checkers can assume val is an 'int' in this branch

Reviewer: Maybe this code block is also not necessary and you could say something like:

> This allows a type checker to narrow the type of ``val`` to ``int`` in the ``else`` block of the example above.

My main concern is readability. Very little has changed in each of these code blocks, but the user needs to scan the whole thing and compare against the previous code block to make sure they're not missing something important.

Author: I was hoping to show the evolution of what the 'else' statement could mean. I'm not sure if I remove any of the examples if that's still clear or not.

Perhaps I can just show this?

```python
if is_str(val)
   # Type checkers can assume val is 'str'
else 
   # This is the only part that would change in the subsequent examples
```

Reviewer: Ok. Thinking about this more, I think having "before this PEP" + "with this PEP" examples makes sense for comparison. But I still think that the first sample is unnecessary. See my other comment.

Since the ``TypeGuard`` type (or output type) is a subtype of the input argument
type, a type checker can determine that the only possible type in the ``else``

> is the other type in the Union. In this example, it is safe to assume that if

Reviewer: ``Union`` or lowercase union.

Author: Oh you mean put it in quotes. Okay.

> For preexisting code this should require no changes, but should simplify this
> use case here:

Reviewer: ```suggestion
For preexisting code this should require no changes, but will allow
use cases such as the one below to be simplified:
```

.. code-block:: python

    class A():
        pass
    class B():
        pass

    def is_A(x: A | B) -> TypeGuard[A]:
        return is_instance(x, A)

    def is_B(x: A | B) -> TypeGuard[B]:
        return is_instance(x, B)

    def test(x: A | B):
        if is_A(x):
            # Do stuff assuming x is an 'A'
            return
        assert is_B(x)

        # Do stuff assuming x is a 'B'
        return

> This use case becomes this instead:

Reviewer: With this proposed change, the code above continues to work but could be simplified by removing the assertion that `x` is of type `B` in the negative case:
